excalidraw/excalidraw
# Syncing optimizations

This changes the sync algorithm to be a lot more efficient. The idea here is each peer keeps track of the elements it has broadcasted and their versions and only broadcasts elements it thinks have changed. Since the server might go down or peers might have connectivity issues we need to assume that messages may be dropped, so we periodically sync all elements (as we do today).

I am not sure the best way to test this w/o doing another giant refactor. Hoping I can crowdsource some testing :)

Test plan:
* Add logging statements to log when a full broadcast happens, partial broadcast happens, and how many elements are synced out.
* Open two windows, draw a few shapes. Reposition and resize the elements. They seem to work.
* Look at the console log. Ensure that the logs show that only 1 element is being synced
* Ensure that a full broadcast happens periodically
* Ensure that a full broadcast happens periodically even while continually dragging a shape for the whole period
* Don't do anything for a while. Full broadcasts stop.
* Start editing again, syncing continues to work and full broadcasts continue.
